File: 2.Routing-in-backend/code/typescript/a_router_in_go_1.ts
// ABSTRACTION: Handler is an interface, it names a capability
// ("be able to handle a request") without saying how. Any class
// implementing handle IS-A Handler. The router depends only on
// this contract, never on a concrete type.
export interface Handler {
    handle(req: Request): Response;
}

export interface Request {
    method: string;
    path: string;
    params: Record<string, string>; // extracted path params (:id)
    query: Record<string, string>;  // query params (?page=2)
}

export interface Response {
    status: number;
    body: string;
}

// ENCAPSULATION: "routes" is private to the class. Outsiders
// cannot touch the table directly; they go through the methods
// we expose (register / dispatch).
export class Router {
    private routes: Map<string, Handler> = new Map(); // key = "METHOD /route"

    // register binds method + route -> handler (the dispatch table).
    public register(method: string, pattern: string, h: Handler): void {
        this.routes.set(`${method} ${pattern}`, h);
    }

    // POLYMORPHISM: dispatch calls h.handle(req) without knowing which
    // concrete type h is. Same call, different behaviour per handler.
    public dispatch(method: string, urlStr: string): Response {
        const [path, queryStr] = urlStr.split('?');
        const query = parseQuery(queryStr || "");
        
        for (const [key, h] of this.routes.entries()) {
            const [m, pattern] = key.split(' ');
            if (m === method) {
                const params = match(pattern, path);
                if (params !== null) {
                    return h.handle({ method, path, params, query }); // polymorphic dispatch
                }
            }
        }
        return { status: 404, body: "route not found" }; // catch-all fallback
    }
}

function parseQuery(queryStr: string): Record<string, string> {
    const query: Record<string, string> = {};
    if (!queryStr) return query;
    for (const pair of queryStr.split('&')) {
        const [k, v] = pair.split('=');
        if (k) query[k] = v || "";
    }
    return query;
}

// match compares "/users/:id" against "/users/123", returning params.
function match(pattern: string, path: string): Record<string, string> | null {
    const p = pattern.replace(/^\/|\/$/g, '').split('/');
    const q = path.replace(/^\/|\/$/g, '').split('/');
    
    if (p.length !== q.length) return null;
    
    const params: Record<string, string> = {};
    for (let i = 0; i < p.length; i++) {
        if (p[i].startsWith(':')) { // dynamic segment
            params[p[i].substring(1)] = q[i]; // bind :id => "123"
        } else if (p[i] !== q[i]) { // static segment must match exactly
            return null;
        }
    }
    return params;
}

// "INHERITANCE" / SHARED BEHAVIOUR: BaseHandler holds shared logic;
// concrete handlers EXTEND it and reuse its methods.
export class BaseHandler implements Handler {
    protected name: string;
    
    constructor(name: string) {
        this.name = name;
    }
    
    protected log(req: Request): void {
        console.log(`[${this.name}] ${req.method} ${req.path}`);
    }
    
    public handle(req: Request): Response {
        throw new Error("Method not implemented.");
    }
}

export class GetUser extends BaseHandler {
    public handle(req: Request): Response {
        this.log(req); // reused (inherited) behaviour
        return { status: 200, body: `user id = ${req.params['id']}` };
    }
}

export class ListBooks extends BaseHandler {
    public handle(req: Request): Response {
        this.log(req);
        let page = req.query['page'];
        if (!page) { page = "1"; }
        return { status: 200, body: `books page ${page}` };
    }
}

if (require.main === module) {
    const r = new Router();
    r.register("GET", "/api/users/:id", new GetUser("users"));
    r.register("GET", "/api/books", new ListBooks("books"));

    console.log(r.dispatch("GET", "/api/users/123").body);   // => user id = 123
    console.log(r.dispatch("GET", "/api/books?page=2").body); // => books page 2
}

File: 2.Routing-in-backend/code/rust/a_router_in_go_1.rs
use std::collections::HashMap;

// ABSTRACTION: Handler is a trait, it names a capability
// ("be able to Handle a request") without saying how. Any type
// implementing Handler IS-A Handler. The router depends only on
// this contract, never on a concrete type.
pub trait Handler {
    fn handle(&self, req: &Request) -> Response;
}

pub struct Request {
    pub method: String,
    pub path: String,
    pub params: HashMap<String, String>, // extracted path params (:id)
    pub query: HashMap<String, String>,  // query params (?page=2)
}

pub struct Response {
    pub status: u16,
    pub body: String,
}

// ENCAPSULATION: "routes" is private to the struct. Outsiders cannot
// touch the table directly; they go through the methods we expose
// (register / dispatch).
pub struct Router {
    // using Box<dyn Handler> for polymorphism
    routes: HashMap<String, Box<dyn Handler>>, // key = "METHOD /route"
}

impl Router {
    pub fn new() -> Self {
        Router {
            routes: HashMap::new(),
        }
    }

    // register binds method + route -> handler (the dispatch table).
    pub fn register(&mut self, method: &str, pattern: &str, h: Box<dyn Handler>) {
        let key = format!("{} {}", method, pattern);
        self.routes.insert(key, h);
    }

    // POLYMORPHISM: dispatch calls h.handle(req) without knowing which
    // concrete type h is. Same call, different behaviour per handler.
    pub fn dispatch(&self, method: &str, url: &str) -> Response {
        let parts: Vec<&str> = url.splitn(2, '?').collect();
        let path = parts[0];
        let query_str = if parts.len() > 1 { parts[1] } else { "" };
        let query = parse_query(query_str);

        for (key, h) in &self.routes {
            let key_parts: Vec<&str> = key.splitn(2, ' ').collect();
            if key_parts.len() != 2 { continue; }
            let r_method = key_parts[0];
            let r_pattern = key_parts[1];

            if r_method == method {
                if let Some(params) = match_pattern(r_pattern, path) {
                    let req = Request {
                        method: method.to_string(),
                        path: path.to_string(),
                        params,
                        query: query.clone(),
                    };
                    return h.handle(&req); // polymorphic dispatch
                }
            }
        }
        Response { status: 404, body: "route not found".to_string() } // catch-all fallback
    }
}

fn parse_query(query_str: &str) -> HashMap<String, String> {
    let mut query = HashMap::new();
    if query_str.is_empty() { return query; }
    for pair in query_str.split('&') {
        let kv: Vec<&str> = pair.splitn(2, '=').collect();
        if kv.len() == 2 {
            query.insert(kv[0].to_string(), kv[1].to_string());
        }
    }
    query
}

// match compares "/users/:id" against "/users/123", filling Params.
fn match_pattern(pattern: &str, path: &str) -> Option<HashMap<String, String>> {
    let p_parts: Vec<&str> = pattern.trim_matches('/').split('/').collect();
    let q_parts: Vec<&str> = path.trim_matches('/').split('/').collect();
    
    if p_parts.len() != q_parts.len() {
        return None;
    }
    
    let mut params = HashMap::new();
    for (p, q) in p_parts.iter().zip(q_parts.iter()) {
        if p.starts_with(':') { // dynamic segment
            let key = &p[1..];
            params.insert(key.to_string(), q.to_string()); // bind :id => "123"
        } else if p != q { // static segment must match exactly
            return None;
        }
    }
    Some(params)
}

// "INHERITANCE" / COMPOSITION: BaseHandler holds shared logic;
// concrete handlers use it for reused behaviour.
pub struct BaseHandler {
    pub name: String,
}

impl BaseHandler {
    pub fn log(&self, req: &Request) {
        println!("[{}] {} {}", self.name, req.method, req.path);
    }
}

pub struct GetUser {
    pub base: BaseHandler,
}

impl Handler for GetUser {
    fn handle(&self, req: &Request) -> Response {
        self.base.log(req); // reused (inherited) behaviour
        let id = req.params.get("id").map(|s| s.as_str()).unwrap_or("unknown");
        Response { status: 200, body: format!("user id = {}", id) }
    }
}

pub struct ListBooks {
    pub base: BaseHandler,
}

impl Handler for ListBooks {
    fn handle(&self, req: &Request) -> Response {
        self.base.log(req);
        let default_page = "1".to_string();
        let page = req.query.get("page").unwrap_or(&default_page);
        Response { status: 200, body: format!("books page {}", page) }
    }
}

fn main() {
    let mut r = Router::new();
    
    r.register("GET", "/api/users/:id", Box::new(GetUser { 
        base: BaseHandler { name: "users".to_string() } 
    }));
    
    r.register("GET", "/api/books", Box::new(ListBooks { 
        base: BaseHandler { name: "books".to_string() } 
    }));

    println!("{}", r.dispatch("GET", "/api/users/123").body);   // => user id = 123
    println!("{}", r.dispatch("GET", "/api/books?page=2").body); // => books page 2
}
